src/poller.rs

```diff
@@ -16,80 +20,317 @@ use crate::{
+where
+    T: 'a + Default + Send,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Poller")
+            .field("name", &self.dbg_name())
+            .field("state", &self.state)
+            .field("interval_us", &self.interval)
+            .field(
+                "thread",
+                &self.thread.map_or_else(
+                    || "<none>".to_string(),
+                    |t| t.name().to_string(),
+                ),
+            )
+            .finish()
+    }
 }
 
-impl<'a, PollerData: 'a> Poller<'a, PollerData> {
-    /// Consumers the poller instance and stops it.
+impl<'a, T> PollerInner<'a, T>
+where
+    T: 'a + Default + Send,
+{
+    fn as_ctx(&self) -> PollerContext {
+        PollerContext(self as *const Self as *mut Self as *mut c_void)
+    }
+
+    fn from_ctx<'b>(p: PollerContext) -> &'b mut Self {
+        unsafe { &mut *(p.0 as *mut Self) }
+    }
+
+    fn dbg_name(&self) -> &str {
+        match &self.name {
+            Some(s) => s.as_str(),
+            None => "<unnamed>",
+        }
+    }
+
+    fn is_active(&self) -> bool {
+        matches!(self.state, PollerState::Running | PollerState::Waiting)
+    }
+
+    fn register(&mut self) {
+        assert_eq!(self.state, PollerState::Starting);
+
+        if let Some(t) = self.thread {
+            info!(
+                "Created an SPDK thread '{}' ({:p}) for poller '{}'",
+                t.name(),
+                t.as_ptr(),
+                self.dbg_name()
+            );
+
+            // Register the poller on its own thread.
+            t.send_msg(self.as_ctx(), |ctx| {
+                let mut p = Self::from_ctx(ctx);
+                p.register_impl();
+            });
+        } else {
+            self.register_impl();
+        }
+    }
+
+    fn register_impl(&mut self) {
+        info!(
+            "Registering new poller '{}' on {}",
+            self.dbg_name(),
+            Thread::current_info(),
+        );
+
+        let poll_fn: spdk_poller_fn = Some(inner_poller_cb::<T>);
+
+        self.inner_ptr = match &self.name {
+            Some(name) => {
+                // SPDK stores the name internally.
+                let name_ptr = name.as_str().into_cstring();
+                unsafe {
+                    spdk_poller_register_named(
+                        poll_fn,
+                        self.as_ctx().0,
+                        self.interval,
+                        name_ptr.as_ptr(),
+                    )
+                }
+            }
+            None => unsafe {
+                spdk_poller_register(poll_fn, self.as_ctx().0, self.interval)
+            },
+        };
+        self.state = PollerState::Waiting;
+    }
+
+    fn stop(&mut self) {
+        info!(
+            "Stopping poller '{}' on {}",
+            self.dbg_name(),
+            Thread::current_info()
+        );
+
+        let _g = self.lock.lock();
+
+        assert_ne!(self.state, PollerState::Stopped);
```

Comment:
Is there never a case where stop is called twice or something like that?

Reply:
No way, it is called from Poller's drop only.
Poller::stop() public method accepts self (not &self), and calls drop, essential it is equivalent to drop.